pkg/metrics/metrics.go

```diff
@@ -0,0 +1,98 @@
+package metrics
+
+import (
+	"encoding/json"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"sigs.k8s.io/prow/pkg/secretutil"
+
+	"github.com/openshift/ci-tools/pkg/api"
+)
+
+// MetricsEvent is the interface that every metric event must implement.
+type MetricsEvent interface {
+	// Store appends the event to the appropriate slice in the MetricsController.
+	Store(mc *MetricsController)
+	// Category returns the event's category.
+	Category() string
+
+	SetTimestamp(time.Time)
+}
+
+const CI_OPERATOR_METRICS_JSON = "ci-operator-metrics.json"
+
+// MetricsController collects and aggregates metrics events.
+type MetricsController struct {
+	events chan MetricsEvent
+	done   chan struct{}
+	wg     sync.WaitGroup
+	mu     sync.Mutex
+
+	insights []InsightsEvent
+	builds   []BuildEvent
+}
+
+// NewMetricsController creates and returns a new MetricsController.
+func NewMetricsController() *MetricsController {
+	return &MetricsController{
+		events: make(chan MetricsEvent),
+		done:   make(chan struct{}),
+	}
+}
+
+// Run listens for events on the events channel until the channel is closed.
+// Once the events channel is closed, we flush the collected events.
+func (mc *MetricsController) Run() {
+	for ev := range mc.events {
```

Comment:
This might race with the `Stop` function. I'd rewrite it in this way
```golang
    mc.wg.Add(1)
    defer mc.wg.Done()
	for ev := range mc.events {
		mc.mu.Lock()
		ev.Store(mc)
		mc.mu.Unlock()
	}
	mc.flush()
```